Trial Banner

nemanjaglumac opened the pull request:
### What does this PR accomplish?
Resolves https://github.com/metabase/metabase/issues/48522.

tl;dr
- We're adding a new app banner that's **only visible and applicable to hosted instances in trial mode**.
    - It should navigate users to the page in store where they can easily manage their subscription.
    - This is a part of the larger "no credit card" sign up initiative

The logic for the banner (re) appearance is as follows:
1. Up until three days before the trial expires it is enough that a user dismisses this banner
2. On the three day before token expiry threshold (mark), the banner re-appears and needs to be dismissed again
3. The same thing happens on the two day mark, one day mark and finally on the 0 day (last 24h) mark

![image](https://github.com/user-attachments/assets/7fea2ef8-2aae-4d41-a969-1681013a6f38)

## Tests
This feature is covered with a lot of unit tests. I plan to add an E2E test in a separate PR. It will involve some mocking and probably playing with the `cy.clock()` so I really don't want to (1) block this PR any longer and (2) increase the number of changed files for now. cc @iethree

nemanjaglumac committed:
Add trial banner

nemanjaglumac commented on frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx:
```diff
@@ -152,4 +158,168 @@ describe("AppBanner", () => {
+    const getTokenExpiryDate = (daysRemaining: number) => {
+      return (
+        dayjs()
+          .add(daysRemaining, "days")
+          // Safety buffer - even a second would do but we want to
+          // account for the slower testing environments, like CI.
+          .add(1, "minute")
```
The alternative would be to mock `now` in order to "freeze the time". Either way is fine with me.

iethree commented on frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:
```diff
@@ -0,0 +1,81 @@
+      icon="warning_round_filled"
+      // This is --mb-base-color-dubloon-10, which is not allowed to be referenced directly.
+      // FIXME: Update the next line once we switch to semantic colors.
+      bg="hsla(46, 96%, 90%, 1)"
```
Yikes! please don't do this. use `--mb-color-warning` (which you can use in mantine with just "warning") - which we want to be uniform throughout the app, and hopefully one day soon will be more dubloony. If this makes your local designer unhappy, send them to me and I will buy them a turtleneck or something.

nemanjaglumac replied:
Done.

https://github.com/metabase/metabase/pull/49344/commits/d24f147c576edcddfe18ce01ec5eecc7887599f4

trinya replied:
@iethree can we make the --mb-color-warning less yellow? where is it used except from this banner? we can have a slack thread if it helps.

nemanjaglumac replied:
We can't.
There is a predefined set of colors currently available in the main app. We absolutely cannot alter the color for a single use case.

The design team came up with a nice color system, but we are currently unable to use it to its full extent. That will change once we upgrade to Mantine v7.

nemanjaglumac committed:
Use the semantic "warning" color for background

iethree commented on frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:
```diff
@@ -0,0 +1,79 @@
+  const lastDay = daysRemaining === 0;
+
+  const shouldShowBanner = useMemo(() => {
+    // No banner if the trial already expired
+    if (daysRemaining < 0) {
+      return false;
+    }
+
+    // In the last 3 days, check dismissal logic with stricter reappearance
+    if (daysRemaining <= 3) {
+      const bannerReappearanceThreshold = tokenExpiryDate
+        .subtract(daysRemaining, "day")
+        .unix();
+
+      // Banner shows if it hasn't been dismissed after the reappearance point
+      return !lastDismissed || lastDismissed < bannerReappearanceThreshold;
+    }
+
+    // For the first 11 days, check if it was ever dismissed
+    return !lastDismissed;
```
The contents of this `useMemo` should be a separate function in a `utils` file with the signature:

```ts
function shouldShowBanner(tokenExpiryDate, lastDismissed)
```

That way you can test nearly all the complex logic in this component with very simple unit tests without the need for mocking settings and whatnot.

iethree commented on frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:
```diff
@@ -0,0 +1,79 @@
+    "trial-banner-dismissal-timestamp",
+  );
+
+  const now = dayjs();
+  const tokenExpiryDate = dayjs(tokenStatus["valid-thru"]);
```
days() is going to get a date from the browser in the user's timezone. I think/guess that our tokens are storing timestamps in UTC? this could create some weird edge cases. We should have tests covering this. Hooray timezones!

nemanjaglumac replied:
This is great observation. It sent me on a long journey, and I thank you for it.
Ironically, it turned out that dayjs automatically handles timezone differences in its methods like `.isSame()`, `.diff()`, etc.

However, I decided to still handle this in the code (1) for the sake of explicitness, and (2) for the sake of easier testing. After all, we might want to change/replace this library with something else in the future. This ensures we don't forget to handle the timezone offset.

Thank you!

iethree commented on frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx:
```diff
@@ -97,8 +100,11 @@ describe("AppBanner", () => {
+        tokenStatus: {
+          ...token,
+          status: "invalid",
+          "error-details": "This is a critical damage.",
```
🙀

iethree commented on frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:
```diff
@@ -0,0 +1,79 @@
+    return !lastDismissed;
+  }, [daysRemaining, lastDismissed, tokenExpiryDate]);
+
+  const href = getStoreUrl("account/manage/plans");
+
+  const handleBannerClose = () => {
+    setLastDismissed(dayjs().unix());
+  };
```
nit, but I would put these below the early return. No need to calculate them if we're not using them.

iethree commented on frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:
```diff
@@ -0,0 +1,79 @@
+      return !lastDismissed || lastDismissed < bannerReappearanceThreshold;
+    }
+
+    // For the first 11 days, check if it was ever dismissed
```
Bug: there is no logic in the code that I can see that prevents this from triggering more than 11 days out. I'm pretty sure in its current state, every trial customer will see a notification that their trial will expire in X days, even if X is 90. The tests also don't cover any number of days greater than 11

nemanjaglumac replied:
No bug, but very vague product documentation.
I double checked with the PM, and the verdict is - as soon as the trial starts (no matter how long it is), show the banner. I'll update tests to make this more explicit.

iethree replied:
🥴

iethree reviewed (changes requested):
Nice job on this feature! ❤️ 

Found one bug, which could really benefit from a little refactoring to make the core date logic easier to test

iethree commented on frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:
```diff
@@ -0,0 +1,79 @@
+
+    // In the last 3 days, check dismissal logic with stricter reappearance
+    if (daysRemaining <= 3) {
+      const bannerReappearanceThreshold = tokenExpiryDate
+        .subtract(daysRemaining, "day")
+        .unix();
```
This is pretty hard to follow. I think what this is saying is that "if the token expires in 3 days, we want to check if the last banner dismissal was before 3 days before the trial expires" ?  🥴 

Maybe instead, let's just check if the banner was dismissed today?


```ts
const wasDismissedToday = dayjs().format("YYYYMMDD") === dayjs(lastDismissed).format("YYYYMMDD");

return !wasDismissedToday;
```

nemanjaglumac replied:
Genial!
I was so deep into this problem/code that I missed the forest for the trees. This is way simpler. Thank you.

nemanjaglumac committed:
Simplify the "reappearance" logic

nemanjaglumac committed:
Refine the banner logic

- Pass the bare minimum information (expiry timestamp) to the `TrialBanner`
- Take timezones into account
- Introduce utilities
- Add unit tests

nemanjaglumac committed:
Update the remaining tests

nemanjaglumac committed:
Clean up

nemanjaglumac committed:
Unbreak my `defsetting` :notes:

nemanjaglumac committed:
Add comments

nemanjaglumac committed:
Defer var assignment

nemanjaglumac committed:
Fix failing tests :facepalm:

nemanjaglumac committed:
Adjust the types in order to match `master`

nemanjaglumac committed:
Move the logic up to the parent component

nemanjaglumac committed:
Convert the `index` component to TS

nemanjaglumac committed:
Fix typo in the JSDoc

nemanjaglumac committed:
Fix another typo

nemanjaglumac committed:
Merge branch 'master' into 48522-trial-banner

iethree commented on frontend/src/metabase/components/AppBanner/AppBanner.tsx:
```diff
@@ -28,6 +40,28 @@ export const AppBanner = () => {
+      lastDismissed,
+    });
+
+    return (
+      showBanner && (
+        <TrialBanner
+          daysRemaining={daysRemaining}
+          onClose={() => setLastDismissed(getCurrentUTCTimestamp())}
+        />
+      )
+    );
```
I think you want to put this return statement in another `if` statement. This logic short-circuits the possibility of showing the payment banner to any trial users.. and is harder to understand than just falling through to the `null` return at the end of the function

```suggestion
    if(showBanner) {
        return (
          <TrialBanner
            daysRemaining={daysRemaining}
            onClose={() => setLastDismissed(getCurrentUTCTimestamp())}
          />
      );
    }

```

nemanjaglumac replied:
Payment banner and Trial banner are mutually exclusive. One is shown to the self-hosted instances and the other one is shown to the hosted/cloud instances.

But I can still reorder this if it makes it easier to read.

iethree reviewed (approved):
Looks good! take a look at these few nits before merging 👇

nemanjaglumac commented on frontend/src/metabase/components/AppBanner/utils.ts:
```diff
@@ -0,0 +1,57 @@
+ * the user has never dismissed it.
+ */
+export function shouldShowTrialBanner({
+  daysRemaining,
```
Honestly, this decision wasn't so much about the performance as much as it was about making this helper pure. I really really **really** didn't want to compute the days remaining inside the function, which makes it dependent on `now` (or the currentTimestamp). Maybe it's just me, but this decision actually made tests much more simple.

If you have two ISO strings as an input, all calculations you do with them are always going to yield the same results. Which also answers to [this comment](https://github.com/metabase/metabase/pull/49344#discussion_r1868408656).

nemanjaglumac commented on frontend/src/metabase/components/AppBanner/utils.ts:
```diff
@@ -0,0 +1,57 @@
+
+  // In the last 3 days, check that the banner was dismissed that day
+  if (daysRemaining <= 3) {
+    const today = dayjs(tokenExpiryTimestamp).subtract(daysRemaining, "days");
```
Please see https://github.com/metabase/metabase/pull/49344#discussion_r1869312211

nemanjaglumac committed:
Make the return statement more readable

nemanjaglumac committed:
Remove redundant test

nemanjaglumac commented:
@metabase-bot backport release-x.51.x
